File: backend/src/models/PledgeCampaign.js
'use strict';

const { Model, DataTypes } = require('sequelize');

const STATUSES = ['draft', 'active', 'closed'];

module.exports = (sequelize) => {
  class PledgeCampaign extends Model {
    static associate(models) {
      PledgeCampaign.hasMany(models.Pledge, { foreignKey: 'campaign_id', as: 'pledges' });
      PledgeCampaign.belongsTo(models.IncomeCategory, {
        foreignKey: 'income_category_id', as: 'incomeCategory'
      });
    }

    get isOpen() {
      return this.status !== 'closed';
    }
  }

  PledgeCampaign.init({
    id: { type: DataTypes.BIGINT, primaryKey: true, autoIncrement: true },
    slug: { type: DataTypes.STRING(64), allowNull: false, unique: true },
    name: { type: DataTypes.STRING(255), allowNull: false },
    name_ti: { type: DataTypes.STRING(255), allowNull: true },
    description: { type: DataTypes.TEXT, allowNull: true },
    description_ti: { type: DataTypes.TEXT, allowNull: true },
    start_date: { type: DataTypes.DATEONLY, allowNull: false },
    end_date: { type: DataTypes.DATEONLY, allowNull: true },
    goal_amount: { type: DataTypes.DECIMAL(10, 2), allowNull: true },
    currency: { type: DataTypes.STRING(3), allowNull: false, defaultValue: 'usd' },
    // VARCHAR + validation rather than a Postgres ENUM: LedgerEntry sets the
    // precedent (enums mapped as STRING to avoid enum mismatch), and it spares
    // us the CREATE TYPE/swap/rename dance on every future status addition.
    status: {
      type: DataTypes.STRING(16),
      allowNull: false,
      defaultValue: 'draft',
      validate: { isIn: [STATUSES] }
    },
    default_payment_type: { type: DataTypes.STRING(50), allowNull: true },
    income_category_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      references: { model: 'income_categories', key: 'id' }
    }
  }, {
    sequelize,
    modelName: 'PledgeCampaign',
    tableName: 'pledge_campaigns',
    underscored: true,
    timestamps: true,
    createdAt: 'created_at',
    updatedAt: 'updated_at'
  });

  PledgeCampaign.STATUSES = STATUSES;

  return PledgeCampaign;
};

File: backendJava/src/main/java/church/abunearegawi/backend/model/PledgeCampaign.java
package church.abunearegawi.backend.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;

/**
 * A fundraising drive. Pledges hang off one of these, and at most one is
 * <em>live</em> at a time.
 *
 * <p>"Live" is not the same as {@code status == active}: it also requires today
 * to fall inside the window. A drive that is active but finished shows nothing
 * to members, which is why the Node backend refuses to activate a campaign
 * whose end date has passed rather than reporting a success that changes
 * nothing. See {@code isLive}.
 *
 * <p>The table is owned by the Node backend's Sequelize migrations; this entity
 * only maps it.
 */
@Entity
@Table(name = "pledge_campaigns")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class PledgeCampaign {

    /** Only ever these three. Stored as a string, not a Postgres enum. */
    public static final String STATUS_DRAFT = "draft";
    public static final String STATUS_ACTIVE = "active";
    public static final String STATUS_CLOSED = "closed";

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 64)
    private String slug;

    @Column(nullable = false)
    private String name;

    @Column(name = "name_ti")
    private String nameTi;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "description_ti", columnDefinition = "TEXT")
    private String descriptionTi;

    @Column(name = "start_date", nullable = false)
    private LocalDate startDate;

    /** Null means open-ended: such a drive never expires. */
    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "goal_amount", precision = 10, scale = 2)
    private BigDecimal goalAmount;

    @Column(nullable = false, length = 3)
    @Builder.Default
    private String currency = "usd";

    @Column(nullable = false, length = 16)
    @Builder.Default
    private String status = STATUS_DRAFT;

    @Column(name = "default_payment_type", length = 50)
    private String defaultPaymentType;

    @Column(name = "income_category_id")
    private Long incomeCategoryId;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private OffsetDateTime updatedAt;

    /** Anything that is not closed still accepts writes. */
    public boolean isOpen() {
        return !STATUS_CLOSED.equals(status);
    }

    /**
     * Whether members can actually see and pledge to this drive right now.
     *
     * <p>Active alone is not enough. A campaign whose window has passed is
     * invisible however its status reads, and treating the two as the same
     * thing is what lets an admin "activate" a finished drive and see no
     * effect anywhere.
     */
    public boolean isLive(LocalDate today) {
        if (!STATUS_ACTIVE.equals(status)) {
            return false;
        }
        if (startDate != null && startDate.isAfter(today)) {
            return false;
        }
        return endDate == null || !endDate.isBefore(today);
    }
}
